libpromises/eval_context.c

```diff
@@ -3847,3 +3900,178 @@ bool EvalContextOverrideImmutableGet(EvalContext *ctx)
     assert(stack_frame != NULL);
     return stack_frame->override_immutable;
 }
+
+// ##############################################################
+
+static EventFrame *EventFrameCreate(char *type, char *name, char *source, SourceOffset offset)
+{
+    EventFrame *event = (EventFrame *) xmalloc(sizeof(EventFrame));
+
+    event->type = type;
+    event->name = name;
+    event->filename = source;
+    event->elapsed = 0;
+    event->children = SeqNew(10, NULL);
+    event->offset = offset;
+    event->parent = NULL;
+
+    char id[strlen(type) + strlen(name) + strlen(source) + PRINTSIZE(offset.start) + PRINTSIZE(offset.line) + 4];
+    xsnprintf(id, sizeof(id)+1, "%s_%s_%s_%ld_%ld", type, name, source, offset.start, offset.line);
+
+    event->id = SafeStringDuplicate(id);
```

Comment:
Use `StringFormat()` instead, so that you don't have to allocate memory on the stack followed by copying onto the heap